File: car_sequencing/CarSequencing.cs
/********** CarSequencing.cs **********/

using System;
using System.IO;
using localsolver;

public class CarSequencing : IDisposable
{
    // Number of vehicles
    int nbPositions;

    // Number of options. 
    int nbOptions;

    // Number of classes. 
    int nbClasses;

    // Options properties. 
    int[] maxCarsPerWindow;
    int[] windowSize;

    // Classes properties. 
    int[] nbCars;
    bool[][] options;

    // Solver. 
    LocalSolver localsolver;

    // LS Program variables. 
    LSExpression[][] classOnPos;

    // Objective.
    LSExpression totalViolations;

    public CarSequencing()
    {
        localsolver = new LocalSolver();
    }

    // Reads instance data.
    void ReadInstance(string fileName)
    {
        using (StreamReader input = new StreamReader(fileName))
        {
            string[] splitted = input.ReadLine().Split(' ');
            nbPositions = int.Parse(splitted[0]);
            nbOptions = int.Parse(splitted[1]);
            nbClasses = int.Parse(splitted[2]);

            splitted = input.ReadLine().Split(' ');

            maxCarsPerWindow = new int[nbOptions];

            for (int o = 0; o < nbOptions; o++)
                maxCarsPerWindow[o] = int.Parse(splitted[o]);

            splitted = input.ReadLine().Split(' ');

            windowSize = new int[nbOptions];

            for (int o = 0; o < nbOptions; o++)
                windowSize[o] = int.Parse(splitted[o]);

            options = new bool[nbClasses][];
            nbCars = new int[nbClasses];

            for (int c = 0; c < nbClasses; c++)
            {
                splitted = input.ReadLine().Split(' ');
                nbCars[c] = int.Parse(splitted[1]);
                options[c] = new bool[nbOptions];
                for (int o = 0; o < nbOptions; o++)
                {
                    int v = int.Parse(splitted[o + 2]);
                    options[c][o] = (v == 1);
                }
            }
        }
    }

    public void Dispose()
    {
        if (localsolver != null)
            localsolver.Dispose();
    }

    void Solve(int limit)
    {
        localsolver = new LocalSolver();

        // Declares the optimization model.
        LSModel model = localsolver.GetModel();

        // classOnPos[c][p] = 1 if class c is at position p, and 0 otherwise
        classOnPos = new LSExpression[nbClasses][];
        for (int c = 0; c < nbClasses; c++)
        {
            classOnPos[c] = new LSExpression[nbPositions];
            for (int p = 0; p < nbPositions; p++)
            {
                classOnPos[c][p] = model.Bool();
            }
        }

        // All cars of class c are assigned to positions
        for (int c = 0; c < nbClasses; c++)
        {
            LSExpression nbCarsFromClass = model.Sum(classOnPos[c]);
            model.Constraint(nbCarsFromClass == nbCars[c]);
        }

        // One car assigned to each position p
        for (int p = 0; p < nbPositions; p++)
        {
            LSExpression nbCarsOnPos = model.Sum();
            for (int c = 0; c < nbClasses; c++)
            {
                nbCarsOnPos.AddOperand(classOnPos[c][p]);
            }
            model.AddConstraint(nbCarsOnPos == 1);
        }

        // optionsOnPos[o][p] = 1 if option o appears at position p, and 0 otherwise
        LSExpression[][] optionsOnPos = new LSExpression[nbOptions][];
        for (int o = 0; o < nbOptions; o++)
        {
            optionsOnPos[o] = new LSExpression[nbPositions];
            for (int p = 0; p < nbPositions; p++)
            {
                optionsOnPos[o][p] = model.Or();
                for (int c = 0; c < nbClasses; c++)
                {
                    if (options[c][o]) optionsOnPos[o][p].AddOperand(classOnPos[c][p]);
                }
            }
        }

        // Number of cars with option o in each window
        LSExpression[][] nbCarsWindows = new LSExpression[nbOptions][];
        for (int o = 0; o < nbOptions; o++)
        {
            nbCarsWindows[o] = new LSExpression[nbPositions - windowSize[o] + 1];
            for (int j = 0; j < nbPositions - windowSize[o] + 1; j++)
            {
                nbCarsWindows[o][j] = model.Sum();
                for (int k = 0; k < windowSize[o]; k++)
                {
                    nbCarsWindows[o][j].AddOperand(optionsOnPos[o][j + k]);
                }
            }
        }

        // Number of violations of option o capacity in each window
        LSExpression[][] nbViolationsWindows = new LSExpression[nbOptions][];
        for (int o = 0; o < nbOptions; o++)
        {
            nbViolationsWindows[o] = new LSExpression[nbPositions - windowSize[o] + 1];
            for (int j = 0; j < nbPositions - windowSize[o] + 1; j++)
            {
                nbViolationsWindows[o][j] = model.Max(0, nbCarsWindows[o][j] - maxCarsPerWindow[o]);
            }
        }

        // Minimize the sum of violations for all options and all windows
        totalViolations = model.Sum();
        for (int o = 0; o < nbOptions; o++)
        {
            totalViolations.AddOperands(nbViolationsWindows[o]);
        }

        model.Minimize(totalViolations);
        model.Close();

        // Parameterizes the solver.
        localsolver.GetParam().SetTimeLimit(limit);

        localsolver.Solve();
    }

    // Writes the solution in a file following the following format: 
    // - 1st line: value of the objective;
    // - 2nd line: for each position p, index of class at positions p.
    void WriteSolution(string fileName)
    {
        using (StreamWriter output = new StreamWriter(fileName))
        {
            output.WriteLine(totalViolations.GetValue());
            for (int p = 0; p < nbPositions; p++)
            {
                for (int c = 0; c < nbClasses; c++)
                {
                    if (classOnPos[c][p].GetValue() == 1)
                    {
                        output.Write(c + " ");
                        break;
                    }
                }
            }
            output.WriteLine();
        }
    }

    public static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: CarSequencing inputFile [outputFile] [timeLimit]");
            Environment.Exit(1);
        }

        string instanceFile = args[0];
        string outputFile = args.Length > 1 ? args[1] : null;
        string strTimeLimit = args.Length > 2 ? args[2] : "60";

        using (CarSequencing model = new CarSequencing())
        {
            model.ReadInstance(instanceFile);
            model.Solve(int.Parse(strTimeLimit));
            if (outputFile != null)
                model.WriteSolution(outputFile);
        }
    }
}

File: car_sequencing/CarSequencing.java
/******** CarSequencing.java *********/

import java.util.*;
import java.io.*;
import localsolver.*;

public class CarSequencing {
    // Number of vehicles.
    private int nbPositions;

    // Number of options.
    private int nbOptions;

    // Number of classes.
    private int nbClasses;

    // Options properties.
    private int[] maxCarsPerWindow;
    private int[] windowSize;

    // Classes properties.
    private int[] nbCars;
    private boolean[][] options;

    // Solver.
    private final LocalSolver localsolver;

    // LS Program variables.
    private LSExpression[][] classOnPos;

    // Objective
    private LSExpression totalViolations;

    private CarSequencing(LocalSolver localsolver) {
        this.localsolver = localsolver;
    }

    // Reads instance data.
    private void readInstance(String fileName) throws IOException {
        try (Scanner input = new Scanner(new File(fileName))) {
            nbPositions = input.nextInt();
            nbOptions = input.nextInt();
            nbClasses = input.nextInt();

            maxCarsPerWindow = new int[nbOptions];
            for (int o = 0; o < nbOptions; o++) {
                maxCarsPerWindow[o] = input.nextInt();
            }

            windowSize = new int[nbOptions];
            for (int o = 0; o < nbOptions; o++) {
                windowSize[o] = input.nextInt();
            }

            options = new boolean[nbClasses][nbOptions];
            nbCars = new int[nbClasses];
            for (int c = 0; c < nbClasses; c++) {
                input.nextInt(); // skip
                nbCars[c] = input.nextInt();
                for (int o = 0; o < nbOptions; o++) {
                    int v = input.nextInt();
                    options[c][o] = (v == 1);
                }
            }
        }
    }

    private void solve(int limit) {
        // Declares the optimization model.
        LSModel model = localsolver.getModel();

        // classOnPos[c][p] = 1 if class c is at position p, and 0 otherwise
        classOnPos = new LSExpression[nbClasses][nbPositions];
        for (int c = 0; c < nbClasses; c++) {
            for (int p = 0; p < nbPositions; p++) {
                classOnPos[c][p] = model.boolVar();
            }
        }

        // All cars of class c are assigned to positions
        for (int c = 0; c < nbClasses; c++) {
            LSExpression nbCarsFromClass = model.sum(classOnPos[c]);
            model.constraint(model.eq(nbCarsFromClass, nbCars[c]));
        }

        // One car assigned to each position p
        for (int p = 0; p < nbPositions; p++) {
            LSExpression nbCarsOnPos = model.sum();
            for (int c = 0; c < nbClasses; c++) {
                nbCarsOnPos.addOperand(classOnPos[c][p]);
            }
            model.constraint(model.eq(nbCarsOnPos, 1));
        }

        // optionsOnPos[o][p] = 1 if option o appears at position p, and 0 otherwise
        LSExpression[][] optionsOnPos = new LSExpression[nbOptions][nbPositions];
        for (int o = 0; o < nbOptions; o++) {
            for (int p = 0; p < nbPositions; p++) {
                optionsOnPos[o][p] = model.or();
                for (int c = 0; c < nbClasses; c++) {
                    if (options[c][o])
                        optionsOnPos[o][p].addOperand(classOnPos[c][p]);
                }
            }
        }

        // Number of cars with option o in each window
        LSExpression[][] nbCarsWindows = new LSExpression[nbOptions][];
        for (int o = 0; o < nbOptions; o++) {
            nbCarsWindows[o] = new LSExpression[nbPositions - windowSize[o] + 1];
            for (int j = 0; j < nbPositions - windowSize[o] + 1; j++) {
                nbCarsWindows[o][j] = model.sum();
                for (int k = 0; k < windowSize[o]; k++) {
                    nbCarsWindows[o][j].addOperand(optionsOnPos[o][j + k]);
                }
            }
        }

        // Number of violations of option o capacity in each window
        LSExpression[][] nbViolationsWindows = new LSExpression[nbOptions][];
        for (int o = 0; o < nbOptions; o++) {
            nbViolationsWindows[o] = new LSExpression[nbPositions - windowSize[o] + 1];
            for (int j = 0; j < nbPositions - windowSize[o] + 1; j++) {
                LSExpression delta = model.sub(nbCarsWindows[o][j], maxCarsPerWindow[o]);
                nbViolationsWindows[o][j] = model.max(0, delta);
            }
        }

        // Minimize the sum of violations for all options and all windows
        totalViolations = model.sum();
        for (int o = 0; o < nbOptions; o++) {
            totalViolations.addOperands(nbViolationsWindows[o]);
        }

        model.minimize(totalViolations);
        model.close();

        // Parameterizes the solver.
        localsolver.getParam().setTimeLimit(limit);

        localsolver.solve();
    }

    // Writes the solution in a file following the following format:
    // - 1st line: value of the objective;
    // - 2nd line: for each position p, index of class at positions p.
    private void writeSolution(String fileName) throws IOException {
        try (PrintWriter output = new PrintWriter(fileName)) {
            output.println(totalViolations.getValue());
            for (int p = 0; p < nbPositions; p++) {
                for (int c = 0; c < nbClasses; c++) {
                    if (classOnPos[c][p].getValue() == 1) {
                        output.print(c + " ");
                        break;
                    }
                }
            }
            output.println();
        }
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: java CarSequencing inputFile [outputFile] [timeLimit]");
            System.exit(1);
        }

        String instanceFile = args[0];
        String outputFile = args.length > 1 ? args[1] : null;
        String strTimeLimit = args.length > 2 ? args[2] : "60";

        try (LocalSolver localsolver = new LocalSolver()) {
            CarSequencing model = new CarSequencing(localsolver);
            model.readInstance(instanceFile);
            model.solve(Integer.parseInt(strTimeLimit));
            if (outputFile != null) {
                model.writeSolution(outputFile);
            }
        } catch (Exception ex) {
            System.err.println(ex);
            ex.printStackTrace();
            System.exit(1);
        }
    }
}
